moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame.cpp

```diff
@@ -448,6 +445,64 @@ void MotionPlanningFrame::importResource(const std::string& path)
         return;
       }
 
+      // If the object is very large, ask the user if the scale should be reduced.
+      bool object_is_very_large = false;
+      for (unsigned int i = 0; i < mesh->vertex_count; i++)
+      {
+        if ((abs(mesh->vertices[i * 3 + 0]) > LARGE_MESH_THRESHOLD) ||
+            (abs(mesh->vertices[i * 3 + 1]) > LARGE_MESH_THRESHOLD) ||
+            (abs(mesh->vertices[i * 3 + 2]) > LARGE_MESH_THRESHOLD))
+        {
+          object_is_very_large = true;
+          break;
+        }
+      }
+      if (object_is_very_large)
+      {
+        QMessageBox msg_box;
+        msg_box.setText(
+            "The object is very large (greater than 10 m). The file may be in millimeters instead of meters.");
+        msg_box.setInformativeText("Attempt to fix the size by shrinking the object?");
+        msg_box.setStandardButtons(QMessageBox::Yes | QMessageBox::No | QMessageBox::Cancel);
+        msg_box.setDefaultButton(QMessageBox::Yes);
+        int ret = msg_box.exec();
+        switch (ret)
+        {
+          case QMessageBox::Yes:
+            // Center the mesh so scaling applies correctly
+            {
+              double sx = 0.0, sy = 0.0, sz = 0.0;
+              for (unsigned int i = 0; i < mesh->vertex_count; ++i)
+              {
+                unsigned int i3 = i * 3;
+                sx += mesh->vertices[i3];
+                sy += mesh->vertices[i3 + 1];
+                sz += mesh->vertices[i3 + 2];
+              }
+              sx /= (double)mesh->vertex_count;
+              sy /= (double)mesh->vertex_count;
+              sz /= (double)mesh->vertex_count;
+              for (unsigned int i = 0; i < mesh->vertex_count; ++i)
+              {
+                unsigned int i3 = i * 3;
+                mesh->vertices[i3] -= sx;
+                mesh->vertices[i3 + 1] -= sy;
+                mesh->vertices[i3 + 2] -= sz;
+              }
```

Comment:
This seems overly complicated: finding the centroid, shifting, scaleAndPadd finding the centroid again, ...
What about just scaling w.r.t zero, i.e. just keeping this code:
```suggestion
              for (unsigned int i = 0; i < mesh->vertex_count; ++i)
              {
                unsigned int i3 = i * 3;
                mesh->vertices[i3] *= 0.001;
                mesh->vertices[i3 + 1] *= 0.001;
                mesh->vertices[i3 + 2] *= 0.001;
              }
```

Reply:
Yes. That's actually what I expected `scaleAndPadd` to do to start with. Might as well keep it simple.